plugins/optimization-detective/optimization.php

```diff
@@ -164,6 +164,11 @@ function od_optimize_template_output_buffer( string $buffer ): string {
 
 	$tag_visitor_registry = new OD_Tag_Visitor_Registry();
 
+	// Send any preload links as Link headers.
+	if ( count( $preload_links ) > 0 && ! headers_sent() ) {
+		header( $preload_links->get_headers() );
+	}
+
```

Comment:
Let's move this down to combine with the existing condition:

```php
	// Inject any preload links at the end of the HEAD.
	if ( count( $preload_links ) > 0 ) {
		$walker->append_head_html( $preload_links->get_html() );
	}
```

For example:

```php
// Send any preload links in a Link response header and in a LINK tag injected at the end of the HEAD.
if ( count( $preload_links ) > 0 ) {
	header( $preload_links->get_response_header() );
	$walker->append_head_html( $preload_links->get_html() );
}
```

Reply:
Hey, so I tried this but encountered errors while running tests.
```bash
➜  wordpress-performance git:(preload-links) npm run test-php
...
There were 9 errors:

1) Test_Image_Prioritizer_Helper::test_image_prioritizer_register_tag_visitors with data set "common-lcp-image-and-lazy-loaded-image-outside-viewport-with-fully-populated-sample-data" (Closure Object (...), '\n                                      <html lang="en">\n                              ...\n                           ', '\n <html lang="en">\n                               ...\n                           ')
Cannot modify header information - headers already sent by (output started at /wordpress-phpunit/includes/bootstrap.php:265)

2) Test_Image_Prioritizer_Helper::test_image_prioritizer_register_tag_visitors with data set "common-lcp-image-with-fully-incomplete-sample-data" (Closure Object (...), '\n                                    <html lang="en">\n                              ...\n                           ', '\n                                  <html lang="en">\n                              ...\n                           ')
Cannot modify header information - headers already sent by (output started at /wordpress-phpunit/includes/bootstrap.php:265)
...
``` 

This is the line in `/wordpress-phpunit/includes/bootstrap.php:265` referenced by the error where the output is started:
```
system( WP_PHP_BINARY . ' ' . escapeshellarg( __DIR__ . '/install.php' ) . ' ' . escapeshellarg( $config_file_path ) . ' ' . $ms_tests . ' ' . $core_tests, $retval );
```

I believe [this section](https://github.com/WordPress/performance/blob/trunk/plugins/optimization-detective/optimization.php#L176-L190) of the `plugins/optimization-detective/optimization.php` is producing some output and `headers()` should be called before any output because after starting to write the http content, headers cannot be sent in the http response as mentioned [here](https://stackoverflow.com/a/8028987/19726929).

Can we move the `$walker->append_head_html( $preload_links->get_html() );` to the above section?
Or is there another solution to this?

Also, should I remove the check for `headers_sent()`?

Reply:
Oh yes, keep the `headers_sent()` check. The issue is that PHPUnit itself is printing output, so it is preventing headers from being sent.

Reply:
Oh, one more important thing: make sure to pass `false` as the second argument to the `header()` function. This will prevent it from overriding any existing `Link` headers that may have been sent.

Reply:
Thanks, will implement these changes.